### PR TITLE
Some usage examples for TensorOptions

### PR DESCRIPTION
**Pull request by ezyang**
Some usage examples for TensorOptions

Pull Request resolved: https://github.com/pytorch/pytorch/pull/11081

Differential Revision: D9579371

### CODE REVIEW

File: aten/src/ATen/TensorOptions.h [hunk 1]
```diff
@@ -17,10 +17,29 @@ namespace at {
 
 /// A class to encapsulate construction axes of a `Tensor`.
 /// `TensorOptions` is a virtual class to enable overriding of certain methods
-/// by subclasses in other libraries, such as PyTorch. In PyTorch, there is a
-/// `torch::TensorOptions` subclass of this `TensorOptions`, which changes
-/// `type()` to return a variable type instead of a tensor type, such that
-/// variables are created inside factory methods, instead of tensors.
+/// by subclasses in other libraries, such as PyTorch.
+///
+/// WARNING: In PyTorch, there are torch:: variants of factory functions,
```
Comment (goldsborough) [inline aten/src/ATen/TensorOptions.h]: This file isn't really about the tensor factories (like `at::ones`), so the comment feels a bit out of place to me, but I guess more documentation doesn't hurt

**Review by goldsborough — APPROVED** [review]

File: aten/src/ATen/TensorOptions.h [hunk 2]
```diff
@@ -17,10 +17,29 @@ namespace at {
 
 /// A class to encapsulate construction axes of a `Tensor`.
 /// `TensorOptions` is a virtual class to enable overriding of certain methods
```
Comment (cpuhrsch) [inline aten/src/ATen/TensorOptions.h]: I'd add some comments about the motivation of 'TensorOptions'. Why is this important?

File: aten/src/ATen/TensorOptions.h [hunk 3]
```diff
@@ -17,10 +17,29 @@ namespace at {
...
-/// variables are created inside factory methods, instead of tensors.
+/// by subclasses in other libraries, such as PyTorch.
+///
+/// WARNING: In PyTorch, there are torch:: variants of factory functions,
+/// e.g., torch::zeros for at::zeros.  These return Variables (while the
+/// stock ATen functions return plain Tensors).  If you mix these functions
+/// up, you WILL BE SAD.
+///
+/// Rather than use the constructor of this class directly, you should prefer to
+/// use the constructor functions, and then chain setter methods on top of them.
```
Comment (cpuhrsch) [inline aten/src/ATen/TensorOptions.h]: 'chain setter methods' is a confusing term to me, but maybe it's common?

File: aten/src/ATen/TensorOptions.h [hunk 4]
```diff
@@ -17,10 +17,29 @@ namespace at {
...
+///
+///     at::device(at::kCUDA).dtype(kInt)
+///     at::dtype(at::kInt)
+///
+/// Additionally, anywhere a TensorOptions is expected, you can directly
+/// pass at::kCUDA/at::kInt, and it will implicitly convert to a TensorOptions.
+///
+/// Here are some common idioms:
```
Comment (cpuhrsch) [inline aten/src/ATen/TensorOptions.h]: What would this look like using TensorOptions? Is it a good idea to add that to showcase the convenience?

File: aten/src/ATen/TensorOptions.h (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/TensorOptions.h]: Method chaining is the standard term for what is going on here: https://en.wikipedia.org/wiki/Method_chaining

File: aten/src/ATen/TensorOptions.h (hunk 4 above)
Comment (ezyang) [reply aten/src/ATen/TensorOptions.h]: These are all using TensorOptions under the hood :) Either by implicit conversions, or functions that return TensorOptions. In ideal usage, you never actually write `TensorOptions`.

File: aten/src/ATen/TensorOptions.h (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/TensorOptions.h]: "The information you need, even when you don't expect it."

File: aten/src/ATen/TensorOptions.h (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/TensorOptions.h]: added

**Pull request closed without merging**